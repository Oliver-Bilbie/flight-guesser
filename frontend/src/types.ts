import { ReactElement } from "react";

export type AlertType = {
  message: string;
  show: boolean;
};

export type ResponseType = {
  message?: string;
  body?: string;
};

export type ResultType = {
  score: number;
  ids: string[];
  message: string;
};

export type FlightData = {
  id: string;
  aircraft: string;
  origin: string;
  destination: string;
  score: string;
};

export type PlayerData = {
  name: string;
  player_id: string;
  score: number;
};

export type SettingsType = {
  useOrigin: boolean;
  useDestination: boolean;
  dataSaver: boolean;
};

export type ButtonType = {
  label: string;
  icon: ReactElement;
  onClick: () => void;
};

export enum LobbyMode {
  hidden,
  join,
  create,
}
