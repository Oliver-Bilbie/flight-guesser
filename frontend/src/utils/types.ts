export type Rules = {
  useOrigin: boolean;
  useDestination: boolean;
};

export type Airport = {
  name: string;
  iata: string;
  lon: number;
  lat: number;
};
