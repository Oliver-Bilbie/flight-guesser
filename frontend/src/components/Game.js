import React, { useState } from "react";
import { Box, Button, Heading, Layer, Spinner, Text } from "grommet";
import { Location } from "grommet-icons";
import AirportSelect from "./AirportSelect";

const Game = () => {
  const [loading, setLoading] = useState(false);
  const [response, setResponse] = useState("");
  const [showResponse, setShowResponse] = useState(false);
  const [airport, setAirport] = useState("");
  const [score, setScore] = useState(0);
  const [ids, setIds] = useState([]);

  const handleSubmit = () => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(handleApi);
    } else {
      setResponse("Geolocation is not supported by this browser.");
      setShowResponse(true);
    }
  };

  const handleResult = (response) => {
    // search for ids which have already been guessed for to prevent duplication
    const duplicate = ids.find((id) => id === response.id);
    if (!duplicate) {
      // only award points for first guesses
      setScore(score + parseInt(response.score));
      setIds(ids.concat(response.id));
    }

    setResponse(
      `You are looking at ${
        ["A", "E", "I", "O", "U"].includes(
          response.aircraft.substring(0, 1).toUpperCase()
        )
          ? "an"
          : "a"
      } ${response.aircraft} aircraft on its way from ${response.origin} to ${
        response.destination
      }.
      You have ${
        duplicate
          ? `already made a guess for this flight`
          : `scored ${response.score} points`
      }.`
    );
  };

  const handleApi = (location) => {
    setLoading(true);
    let request = new XMLHttpRequest();
    const path = process.env.REACT_APP_ENDPOINT;
    const body = `{"x": ${location.coords.longitude}, "y": ${location.coords.latitude}, "airport": "${airport}"}`;

    request.onerror = function () {
      setResponse("An error has occurred");
      setLoading(false);
      setShowResponse(true);
    };
    request.ontimeout = function () {
      setResponse("The request has timed out");
      setLoading(false);
      setShowResponse(true);
    };
    request.onload = function () {
      if (request.status === 200) {
        if (request.response.status === 200) {
          handleResult(request.response.response);
        } else if (request.response.status === 400) {
          setResponse("No nearby flights were found");
        } else {
          setResponse("Data for the nearby flight is not available");
        }
      } else {
        setResponse("Unable to fetch data");
      }
      setLoading(false);
      setShowResponse(true);
    };

    request.timeout = 30000;
    request.responseType = "json";
    request.open("POST", path);
    request.send(body);
  };

  return (
    <Box direction="row" align="center">
      {loading ? (
        <Spinner size="large" />
      ) : (
        <Box gap="medium">
          <Heading textAlign="center">Score: {score}</Heading>
          <Text>Destination:</Text>
          <AirportSelect handleSelection={setAirport} />
          <Box width="190px" alignSelf="center" pad={{vertical: "medium"}}>
          <Button
            label="Make Guess"
            icon={<Location />}
            onClick={handleSubmit}
          />
          </Box>
        </Box>
      )}
      {showResponse && (
        <Layer
          onEsc={() => setShowResponse(false)}
          onClickOutside={() => setShowResponse(false)}
        >
          <Box
            width="medium"
            pad="small"
            gap="medium"
            align="center"
            justify="center"
          >
            <Text>{response}</Text>
            <Box width="xsmall">
              <Button label="Close" onClick={() => setShowResponse(false)} />
            </Box>
          </Box>
        </Layer>
      )}
    </Box>
  );
};

export default Game;
