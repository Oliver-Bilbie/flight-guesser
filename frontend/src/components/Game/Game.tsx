import React, { useState, useEffect } from "react";
import { Box, Button, Heading, Layer, Spinner, Text } from "grommet";
import { Location, Performance } from "grommet-icons";
import AirportSelect from "../AirportSelect/AirportSelect";
import SettingsMenu from "../SettingsMenu/SettingsMenu";
import { handleTurnApi, handleResult } from "../../helpers/handle_turn";
import { getAirportsApi } from "../../helpers/get_airports";

const Game: React.FC = (): React.ReactElement => {
  const [settingsValues, setSettingsValues] = useState({
    useOrigin: false,
    useDestination: true,
    dataSaver: true,
  });
  const [showSettings, setShowSettings] = useState(false);
  const [loading, setLoading] = useState(true);
  const [response, setResponse] = useState("");
  const [showResponse, setShowResponse] = useState(false);
  const [guess, setGuess] = useState({ origin: "", destination: "" });
  const [score, setScore] = useState(0);
  const [ids, setIds] = useState([] as string[]);
  const [airports, setAirports] = useState([] as string[]);

  useEffect(() => {
    getAirportsApi(setAirports);
    setLoading(false);
  }, []);

  const handleSubmit = (): void => {
    if (navigator.geolocation) {
      if (
        (guess.origin !== "" || !settingsValues.useOrigin) &&
        (guess.destination !== "" || !settingsValues.useDestination)
      ) {
        setLoading(true);
        navigator.geolocation.getCurrentPosition((location) =>
          handleTurnApi(
            location.coords.longitude,
            location.coords.latitude,
            guess.origin,
            handleTurn
          )
        );
      } else {
        setResponse("Please select an airport");
        setShowResponse(true);
      }
    } else {
      setResponse("Location services must be enabled to use this application");
      setShowResponse(true);
    }
  };

  const handleTurn = (response): void => {
    if (response.message) {
      setResponse(response.message);
    } else {
      const result = handleResult(response.body, score, ids);
      setResponse(result.message);
      setScore(result.score);
      setIds(result.ids);
    }
    setLoading(false);
    setShowResponse(true);
    setGuess({ origin: "", destination: "" });
  };

  return (
    <>
      <Box
        direction="column"
        align="center"
        width="large"
        background="light-2"
        elevation="small"
        margin="small"
        round
      >
        {loading ? (
          <Spinner size="large" />
        ) : (
          <Box gap="medium">
            <Heading textAlign="center">Score: {score}</Heading>
            <AirportSelect
              airports={airports}
              setSelection={(selection: string): void =>
                setGuess({ ...guess, origin: selection })
              }
            />
            <Box width="190px" alignSelf="center" pad={{ vertical: "medium" }}>
              <Button
                label="Make Guess"
                icon={<Location />}
                onClick={handleSubmit}
              />
            </Box>
          </Box>
        )}
        <Button
          icon={<Performance />}
          onClick={(): void => setShowSettings(!showSettings)}
          alignSelf="end"
          hoverIndicator
        />
      </Box>
      {showResponse && (
        <Layer
          onEsc={(): void => setShowResponse(false)}
          onClickOutside={(): void => setShowResponse(false)}
        >
          <Box
            width="medium"
            pad="small"
            gap="medium"
            align="center"
            justify="center"
          >
            <Text>{response}</Text>
            <Box width="xsmall">
              <Button
                label="Close"
                onClick={(): void => setShowResponse(false)}
              />
            </Box>
          </Box>
        </Layer>
      )}
      {showSettings && (
        <SettingsMenu
          settingsValues={settingsValues}
          setSettingsValues={setSettingsValues}
          onClose={(): void => setShowSettings(false)}
        />
      )}
    </>
  );
};

export default Game;
