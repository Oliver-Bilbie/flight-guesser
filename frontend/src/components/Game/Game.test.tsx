import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import Game from "./Game";
import { ResultType } from "../../types";

// Mock API calls
jest.mock("../../helpers/get_airports", () => ({
  getAirportsApi(): string[] {
    return ["airport 1", "airport 2", "airport 3"];
  },
}));

jest.mock("../../helpers/handle_turn", () => ({
  handleTurnApi(): void {},
  handleResult(): ResultType {
    return {
      message: "test message",
      score: 10,
      ids: ["id1", "id2", "id3"],
    };
  },
}));

// Mock GeoLocation
const mockGeolocation = {
  getCurrentPosition: jest.fn(),
  watchPosition: jest.fn(),
};
// @ts-ignore
global.navigator.geolocation = mockGeolocation;

afterEach(cleanup);

test("Test with no guess", () => {
  render(<Game />);

  // Check that the expected components are rendered
  expect(screen.getByText("Score: 0")).toBeInTheDocument();
  expect(screen.getByText("Make Guess")).toBeInTheDocument();

  // Click on the "Make Guess" button and check that the expected components render
  fireEvent.click(screen.getByText("Make Guess"));
  expect(
    screen.getByText("Please select a destination airport")
  ).toBeInTheDocument();
  expect(screen.getByText("Close")).toBeInTheDocument();
});

// TODO: Write tests where an airport is selected
