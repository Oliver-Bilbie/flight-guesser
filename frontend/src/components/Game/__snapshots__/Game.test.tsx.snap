// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 jvmpif"
>
  <div
    className="StyledBox-sc-13pk1d4-0 gmiLjI"
  />
  <button
    className="StyledButton-sc-323bzc-0 hiNAcw"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    <svg
      aria-label="Performance"
      className="StyledIcon-sc-ofa7kd-0 kKhdEN"
      viewBox="0 0 24 24"
    >
      <path
        d="m12 19-2 3-3-1-.5-3.5L3 17l-1-3 3-2-3-2 1-3 3.5-.5L7 3l3-1 2 3 2-3 3 1 .5 3.5L21 7l1 3-3 2 3 2-1 3-3.5.5L17 21l-3 1-2-3zm0-3a4 4 0 1 0 0-8 4 4 0 0 0 0 8z"
        fill="none"
        stroke="#000"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
  </button>
</div>
`;
